### ISSUE TITLE
Forward material coupled variable dependencies to kernels

### ISSUE THREAD

**Issue by GiudGiud** [idaholab/moose#25800]
## Reason
Reported here https://github.com/idaholab/moose/discussions/25796#discussioncomment-7343259
Kernels seemingly (is that even documented?? I dont see it) rely on the addCoupledVar to decide which off-diagonals to compute when not using AD.

If a user is using materials to compute variable dependent terms, they ll miss those calls. This is a pretty big oversight if it's true (I have not checked)

see this page for the lack of docs
https://mooseframework.inl.gov/source/interfaces/Coupleable.html

## Design
Investigate the extent of the problem 
In the Matpropinterface that the kernel inherits, add some special logic to forward the dependencies

## Impact
Less silent failures

**Comment by joshuahansel** [comment]
Here's a minimum example confirming the behavior: [matdep.txt](https://github.com/user-attachments/files/26055617/matdep.txt). In this example, there's no implementation of hand-coded Jacobians corresponding to the material property, but as I just noted in the linked discussion, getting the `jvar` that you would use in the hand-coded off-diagonal Jacobian for the kernel requires that you couple the variable anyway.

**Comment by joshuahansel** [comment]
So to summarize, this is my understanding:
- If someone made an attempt to compute the Jacobian for a coupled variable, that's already being used if you assume that they have called `coupled("myvar")` to get the `jvar` (else, where did they get it?)
- Users can make mistakes by forgetting to code a Jacobian entry due to a material. If we ensure we call `computeOffDiagonalJacobian(jvar)` for the dependencies of the material, nothing actually happens because they forgot to code it. The best we could do is give an error/warning if they forgot to code it, right? If the coupling entries for the kernel does not include all of the material variable dependencies, then give warning/error? This seems a bit iffy, since we can't truly get material *property* dependencies, only *material* dependencies: a material may contain several properties, which depend on various sets of the variables coupled to the material.

**Comment by GiudGiud** [comment]
> that's already being used if you assume that they have called coupled("myvar") to get the jvar (else, where did they get it?)

Right. If you assume they coupled the variable in the kernel to get the index, then clearly they have manually forwarded the dependence and they are in the clear.

> This seems a bit iffy, since we can't truly get material property dependencies, only material dependencies:

That's fair. We could use a Jacobian finite difference check to determine if a kernel ended up being affected. But ideally we'd have something in between to do that
